### PR TITLE
wip: thread local context information

### PR DESCRIPTION
**Pull request by bluejekyll**

### CODE REVIEW

**Comment by bluejekyll** [comment]
@jamessewell, this is that branch I was playing with. Right now it only enables use of the pg_allocator in the primary postgres thread, other threads will use the system allocator.

**Comment by jamessewell** [comment]
Huh I didn't know you *could* do that! I'll have a look.

**Comment by bluejekyll** [comment]
Maybe you can’t. This is not stable.

File: pg-extend/src/pg_alloc.rs [hunk 1]
```diff
@@ -7,23 +7,89 @@
...
+        IN_PGCONTEXT.with(Cell::get)
+    }
+
 unsafe impl GlobalAlloc for PgAllocator {
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        // TODO: is there anything we need ot do in terms of layout, etc?
-        pg_sys::palloc(layout.size()) as *mut u8
+        // PgAllocator::init();
+
+        if IN_PGCONTEXT.with(Cell::get) {
```
Comment (intgr) [inline pg-extend/src/pg_alloc.rs]: Uhh, this looks incredibly unsafe. If `palloc` gets used to allocate in the main thread, then that variable gets transferred to another Rust thread which tries to free it, that would crash.

I think a better idea would be to disable `PgAllocator` in threaded Rust extensions entirely; when returning results from a Rust function, that data can be copied into a `palloc`-allocated buffer.

Comment (bluejekyll) [reply pg-extend/src/pg_alloc.rs]: Agreed. This entire idea is unsafe.

**Comment by bluejekyll** [comment]
FYI, I'm closing this. At this point I'm not convinced there is a way to safely use the Postgres allocator in Rust. I'm thinking that we have to be really strict about the boundaries of Postgres' C interfaces and Rust's.

Let rust's allocator allocate for Rust, and Postgres' allocator allocate for PG. This will mean that any data that's not stack based, will need to be copied into memory allocated by Postgres. On the Rust side, we can possibly impose some strict lifetimes on memory passed into Rust that can make it safe to reference that data.

**Pull request closed without merging**